### ISSUE TITLE
CA2021 false positive on Enumerable.OfType<T>extension

### ISSUE THREAD

**Issue by cizmiak** [dotnet/roslyn-analyzers#7153]
### Analyzer
**Diagnostic ID**: [CA2021](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2021): `CA2021: Don't call Enumerable.Cast<T> or Enumerable.OfType<T> with incompatible types`

### Analyzer source
**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)
**Version**: [SDK 8.0.101](https://dotnet.microsoft.com/download/dotnet/8.0)

### Describe the bug
Warning CA2021 is found on Enumerable.OfType<T> call when T is compatible with underlining type in the enumerable collection.

### Steps To Reproduce
```
	public static void Main()
	{
		var nodeChanges = new List<INodeUpdate<GraphNode>> { new DataNodeUpdate(new DataNode(0, 0)) };

		//warning CA2021: This call will always result in an empty sequence because type 'INodeUpdate<GraphNode>' is incompatible with type 'DataNodeUpdate'
		var nodeChangesFiltered = nodeChanges.OfType<DataNodeUpdate>();

		Debug.Assert(nodeChangesFiltered.Count() == 1);
	}

	public abstract record class GraphNode(long Id);
	public sealed record class DataNode(long Id, long Value) : GraphNode(Id);

	public interface INodeUpdate<out T>
	{
		T Updated { get; }
	}

	public abstract record class NodeUpdate<T>(T Updated) : INodeUpdate<T> where T : GraphNode;
	public sealed record class DataNodeUpdate(DataNode Updated) : NodeUpdate<DataNode>(Updated);
```

### Expected behavior
No CA2021 warning.

### Actual behavior
CA2021 warning is found.

**Comment by mavasani** [comment]
Tagging @buyaa-n - is this related to #6457? Are the fixes in that PR part of .NET 8 SDK?

**Comment by buyaa-n** [comment]
Doesn't look like it is fixed with https://github.com/dotnet/roslyn-analyzers/issues/6457, @fowl2 could you take a look?
